src/lib/rpc/bestRpcTracker.ts

```diff
@@ -0,0 +1,368 @@
+
+// DataStore field used for probing
+const PROBE_SAMPLE_FIELD = "minCollateralFactor";
+// Markets used for `PROBE_SAMPLE_FIELD` reading
+const PROBE_SAMPLE_MARKET = {
+  [ARBITRUM]: "0x70d95587d40A2caf56bd97485aB3Eec10Bee6336", // ETH/USD
+  [AVALANCHE]: "0xB7e69749E3d2EDd90ea59A4932EFEa2D41E245d7", // ETH/USD
+  [AVALANCHE_FUJI]: "0xbf338a6C595f06B7Cfff2FA8c958d49201466374", // ETH/USD
+};
+
+type RpcProbe = {
+  isSuccess: boolean;
+  responseTime: number;
+  blockNumber: number | null;
+  timestamp: Date;
+};
+
+type ProviderData = {
+  url: string;
+  provider: Provider;
+  probeData: RpcProbe[];
+};
+
+type RpcTrackerState = {
+  [chainId: number]: {
+    lastUsage: Date;
+    currentBestProviderUrl: string;
+    providers: {
+      [providerUrl: string]: ProviderData;
+    };
+  };
+};
+
+let trackingIntervalId: NodeJS.Timeout | null = null;
+let trackerState: RpcTrackerState | null = null;
+
+function initRpcTracking() {
+  trackerState = initTrackerState();
+
+  if (trackingIntervalId) {
+    clearInterval(trackingIntervalId);
+  }
+  measureRpcData();
+  trackingIntervalId = setInterval(measureRpcData, PROBE_INTERVAL);
+}
+
+function measureRpcData() {
+  if (!trackerState) {
+    throw new Error("RPC tracker state is not initialized");
+  }
+
+  entries(trackerState).forEach(async ([chainIdRaw, chainTracker]) => {
+    const chainId = Number(chainIdRaw);
+    const providers = Object.values(chainTracker.providers);
+
+    if (
+      differenceInMilliseconds(Date.now(), chainTracker.lastUsage) > DISABLE_UNUSED_TRACKING_TIMEOUT ||
+      providers.length < 2
+    ) {
+      return;
+    }
+
+    const probes = providers.map(async (providerInfo) => {
+      providerInfo.probeData = providerInfo.probeData.filter((probe) => {
+        return Date.now() - probe.timestamp.getTime() < PROBE_SLIDING_WINDOW;
+      });
+
+      return probeRpc(chainId, providerInfo.provider, providerInfo.url).then((probeResult) => {
+        providerInfo.probeData.push(probeResult);
+
+        return probeResult;
+      });
+    });
+
+    const probeResults = await Promise.all(probes);
+    const successProbeResults = probeResults.filter((probe) => probe.isSuccess);
+
+    if (!successProbeResults.length) {
+      const fallbackUrl = getBestRpc(chainId).fallback;
+
+      if (fallbackUrl) {
+        setCurrentProvider(chainId, fallbackUrl);
+      }
+
+      return;
+    }
+
+    const probeResultsByBlockNumber = orderBy(successProbeResults, ["blockNumber"], ["desc"]);
+    let bestBlockNumberProbe = probeResultsByBlockNumber[0];
+
+    const probeStats = probeResultsByBlockNumber.map((probe, i, arr) => {
+      let isValid = probe.isSuccess;
+
+      const bestBlockNumber = bestBlockNumberProbe?.blockNumber;
+      const currProbeBlockNumber = probe.blockNumber;
+      const nextProbeBlockNumber = arr[i + 1]?.blockNumber;
+
+      // Rare case when RPC returned a block number from the far future
+      if (
+        i === 0 &&
```

Comment:
imagine same provider holding 3 top places in successProbeResults, that way you would challenge provider vs itself. you need to determine bestBlockNumberProbe separately outside of this loop

Reply:
Could you elaborate on this thought? I don't understand how one provider can take several places when different providers compete with each other in design

Reply:
tbh i personally don't like to use adjacent array items inside the filter/map loop. i think it's less readable. imo something instead of current loop like this would be simpler:

```
// remove item with block from future
if (
  probeResultsByBlockNumber.length >= 2 &&
  probeResultsByBlockNumber[0].blockNumber - probeResultsByBlockNumber[1].blockNumber > BLOCK_FROM_FUTURE_THRESHOLD
) {
  probeResultsByBlockNumber.shift();
}
const bestBlockNumber = probeResultsByBlockNumber[0].blockNumber;

// filter out items with lagging blocks
probeResultsByBlockNumber = probeResultsByBlockNumber.filter(p => bestBlockNumber - p.blockNumber <= BLOCK_LAGGING_THRESHOLD);
```

unless i'm missing something, there is no need to create probeStats. you just need to filter out "bad" results

Reply:
I also don't like this place, but the reason to keep this cycle - is to get the probeStats array which is used to show the debug table with probe data in debug mode.

If you'd like it more, we can do something like that: 

```
let bestBlockNumberProbe = probeResultsByBlockNumber[0];
const nextProbeBlockNumber = probeResultsByBlockNumber[1]?.blockNumber;

if (
  bestBlockNumberProbe?.blockNumber &&
  nextProbeBlockNumber &&
  bestBlockNumberProbe.blockNumber - nextProbeBlockNumber > BLOCK_FROM_FUTURE_THRESHOLD
) {
  bestBlockNumberProbe = probeResultsByBlockNumber[1];
}

const probeStats = probeResultsByBlockNumber.map((probe, i) => {
  let isValid = true;

  const bestBlockNumber = bestBlockNumberProbe.blockNumber;
  const currProbeBlockNumber = probe.blockNumber;

  if (bestBlockNumber && currProbeBlockNumber && bestBlockNumber - currProbeBlockNumber > BLOCK_LAGGING_THRESHOLD) {
    isValid = false;
  }

  return {
    ...probe,
    isValid,
  };
});
```

Reply:
I vote to move this logic outside of a loop since we only want to check only first provider

btw don't we want to deal with cases when 2 or more providers are from future? I thought logic could be - find the most "correct" block

now we're fixing case: 1 2 3 4 5 100
what if we have case: 1 2 3 4 5 100 100?

maybe logic should prioritize majority and punish extereme difference?

Reply:
1. Updated this place. Now the "block from the future" case handling is outside from the cycle
2. Discussed, that it's nearly impossible that the 1 2 3 4 5 100 100 case can happen